File: dynamic_dns.pl
#!/usr/bin/perl

# use strict;
use YAML::XS qw(LoadFile DumpFile);
use HTTP::Tiny;
use Email::Send;
use Email::Send::Gmail;
use Email::Simple::Creator;

my $config_file  = "dynamic_dns.yml";
my $log_filename = "dynamic_dns.log";

open(my $log_fh, '>>', $log_filename) or die "Cannot open log file ($log_filename) for appending.\n$!\n";

my $config = LoadFile($config_file);

sub output {
    my $action = shift;
    my $message = shift;
    my $out = sprintf("%04d-%02d-%02d %02d:%02d:%02d: %s\n", (localtime)[5]+1900, (localtime)[4]+1, (localtime)[3,2,1,0], sprintf($message, @_));

    print $log_fh $out;

    if($action eq 'email' or $action eq 'warn' or $action eq 'die') {
        my $email = Email::Simple->create(
            header => [
                From    => $config->{config}->{email_from},
                To      => $config->{config}->{email_to},
                Subject => sprintf("Dynamic DNS Error: %s", $action),
                ],
            body => sprintf("Dynamic DNS Script %s.\n\n%s.", $action, $out)
            );

        my $sender = Email::Send->new(
            {   mailer      => 'Gmail',
                mailer_args => [
                    username => $config->{config}->{username},
                    password => $config->{config}->{password},
                ]
            }
        );
        $sender->send($email) or warn "Error sending email: $@";
    }

    if($action eq 'warn') { warn $out; }
    elsif($action eq 'die') { die $out; }
    else { print $out; }

    return;
}

foreach my $site (sort keys %{$config->{sites}}) {
    next if($site eq 'config');
    next if($config->{sites}->{$site}->{last_result} eq 'error');

    my $url = sprintf('https://%s:%s@domains.google.com/nic/update?hostname=%s', 
                      $config->{sites}->{$site}->{username}, 
                      $config->{sites}->{$site}->{password},
                      $site);
    if($config->{sites}->{$site}->{use_local_ip}) {
        my $interface = $config->{sites}->{$site}->{use_local_ip};
        my $local_ip = `ip -4 address show $interface`;
        $local_ip =~ /inet ([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})/;
        $url .= "\&myip=$1";
    }
    my $response = HTTP::Tiny->new->get($url);

    output('warn', 'DNS API call failed.') unless $response->{success};

    if($response->{content} =~ /^(good|nochg)/) {
        output('print', "%s: %s", $site, $response->{content});
    } else {
        $config->{sites}->{$site}->{last_result} = 'error';
        output('warn', "%s: %s", $site, $response->{content});
    }
}

DumpFile($config_file, $config);

File: dynamic_dns.py
#!/usr/bin/python3

import logging
import os
import re
import requests
import smtplib
import socket
import ssl
import yaml

my_hostname = socket.gethostname()
CONFIG_FILENAME = os.path.join(os.path.dirname(os.path.abspath(sys.argv[0])),
                               f"dynamic_dns.{my_hostname}.yml")

config = []
with open(CONFIG_FILENAME, 'r') as config_file:
    try:
        config = yaml.safe_load(config_file)
    except yaml.YAMLError as exc:
        print(exc)

logging.basicConfig(filename=config.get('config', {}).get('log_filename'),
                    encoding='utf-8',
                    level=logging.INFO,
                    format='%(asctime)s %(message)s')
logger = logging.getLogger()


def get_ip():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        # doesn't even have to be reachable
        s.connect(('10.255.255.255', 1))
        IP = s.getsockname()[0]
    except Exception:
        IP = '127.0.0.1'
    finally:
        s.close()
    return IP


def get_ip6():
    s = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
    try:
        s.connect(('2001:4860:4860::8888', 53))
        IP = s.getsockname()[0]
    except Exception:
        IP = '::1'
    finally:
        s.close()
    return IP


def get_public_ip(api_url):
    result = requests.get(api_url)
    if result.status_code != requests.codes.ok:
        message = f'IP API call failed. {result.text}'
        logging.error(message)
        email(message)
        return None
    else:
        return result.text


def email(message):
    email_text = f"""\
From: {config['config']['email_from']}
To: {config['config']['email_to']}
Subject: Dynamic DNS error: {my_hostname}

{message}
"""

    try:
        # Create a secure SSL context
        context = ssl.create_default_context()

        with smtplib.SMTP_SSL('smtp.gmail.com', 465, context=context) as server:
            server.login(config['config']['username'],
                         config['config']['password'])
            server.sendmail(config['config']['email_from'],
                            config['config']['email_to'], email_text)
            server.close()

        logger.warning('Email sent to %s.', config['config']['email_to'])
    except:
        logger.error('Something went wrong while sending the failure email.')


for site in config['sites']:
    if not (('last_result' in config['sites'][site]) and (config['sites'][site]['last_result'] != 'error')):
        payload = {}
        payload['hostname'] = site
        if 'use_local_ip' in config['sites'][site]:
            payload['myip'] = get_ip()
        if 'use_local_ip6' in config['sites'][site]:
            payload['myip'] = get_ip6()
        if 'ip_api_url' in config['sites'][site]:
            payload['myip'] = get_public_ip(
                config['sites'][site]['ip_api_url'])

        result = requests.get(config['config']['api_url'], auth=(
            config['sites'][site]['username'], config['sites'][site]['password']), params=payload)
        logger.info('%s: %s', site, result.text)

        if (result.status_code != requests.codes.ok) or not (re.match('(good|nochg)', result.text)):
            logger.error(f'DNS API call failed for {site}. {result.text}')
            email(f'DNS API call failed for {site}. {result.text}')
            config['sites'][site]['last_result'] = 'error'

with open(CONFIG_FILENAME, 'w') as config_file:
    yaml.dump(config, config_file, default_flow_style=False)
